### PR TITLE
Improve handling of addon config changes on kube upgrade

### LINKED ISSUE #5837: Improve handling of addon config changes on kube upgrade
SURE-8310
SURE-8932

**Setup**
- Rancher version: 2.6.4

**Describe the bug**
For background see https://github.com/rancher/dashboard/issues/5672. 

Overview
- RKE2 clusters have addons. The version of an addon is tied to the version of kubernetes. Addons have their own configuration
- I believe addons are helm chart based and their configuration is based on the usual helm values process (there is a default set that can be granulary overwritten by users)
- Everything is pretty straight forward when an RKE2 cluster is created
  - User selects a kube version and addon. 
  - We show the default yaml for the addon version and allow the user to change it. The diff is cached by addon version
  - If a user changes the addon version (by switching to a different addon or kube version) we show the new default yaml overlaid with their cached copy
- The issue occurs when the user edits the cluster and changes the addon version (primarily by changing the kube version)
  - We show the default values yaml without applying the config from the previous version. This means user has to re-configure the addon. If they don't the edit may fail
  - As a workaround when the kube version is changed we show a warning modal that their config may need updating

Objections were raised about this workaround in https://github.com/rancher/dashboard/issues/5672#issuecomment-1113647204

Ideally we should avoid the scenario where a user has to re-enter their configuration

We need to understand
- How addon config works, specifically confirm if it is helm values based
- How addons are applied on cluster edits with a kube upgrade. Is this a fresh helm install, or an upgrade

Given above we might be able to apply the users config (if it is still a diff) on top of the new addon versions yaml (I think this matches how a normal helm upgrade works). Otherwise we're in a tricky scenario of trying to match properties

There's some detail about how addon / addon config works in our confluence (`Container UI/UX / Projects / Rancher / Cluster Provisioning` / `General Notes`)

**To Reproduce**
To supply when this is tackled, but basically the RKE2 create / edit page addon tab

**Result**

**Expected Result**

**Screenshots**
image

image

### PR DESCRIPTION
**Pull request by momesgin**
### Summary
Fixes #5837 

### Occurred changes and/or fixed issues

This PR stops unnecessary confirmation dialog from showing up when there's no conflict between the user's custom values and the diffeneces between the addon’s old and new default values. We now re-apply the user's old addon values when there's no conflict.

### Technical notes summary

- Addon config preservation logic:
  - Added new logic in `rke2.vue` to smartly handle addon configs when a user updates the Kubernetes version while editing a cluster.
  - When the K8s version changes, we now compare the default addon values between the old and new versions.
  - A difference only counts as relevant if it’s something the user actually customized. It gets determined by checking `userChartValues`.
  - The logic only runs for addons that are actually enabled in the cluster(i.e. from `this.addonNames`).
  - The new logic is separted into multiple helper methods (`_processAddonVersionChange` and `_filterRelevantDifferences`)
   - Changed the watcher for the `selectedVersion` computed property from a simple method to an object with an async handler, so it can properly handle the new async addon logic.
   - Added a new `addonConfigDiffs` data property to track relevant differences. It’s used to decide if we need to show the confirmation modal.
- Unit Testing:
  - Added a new tests for the addon preservation feature in `rke2.test.ts`.
  - Some unrelated tests were cleaned up to fix console warnings.

### Areas or cases that should be tested

**IMPORTANT**: I tested this for vSphere CPI/CNI addons but if you know other addons that can be tested for this behavior please test them instead.

Create a cluster and choose an old version of Kubernetes, then test these scenarios:

- Add no changes to your testing addon, then update the kubernetes version(try multiple version)
- Add some changes to your testing addon configuration, then update the kubernetes version to a version that's not requiring a newer version of your addon
- Add some changes to your testing addon configuration, then update the kubernetes version to a version that requires a newer version of your addon

### Areas which could experience regressions

Since the changes are in the main RKE2 cluster edit component (`rke2.vue`), a quick sanity check of saving a cluster after changing other, unrelated values (like labels or description) would be nice.

A quick check should be done to ensure that switching to the YAML view and making an edit there doesn't interfere with this new logic, although it's unlikely as the changes are primarily in the watcher for the K8s version dropdown.

### Screenshot/Video

Before:

https://github.com/user-attachments/assets/1a4c406b-ce39-4e78-aefe-62930201190a

After:

https://github.com/user-attachments/assets/b3b1d14a-81ca-4d57-ab5f-9da50253c4e5

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone 
- [x] The PR template has been filled out
- [x] The PR has been self reviewed 
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes
- [x] The PR has been reviewed in terms of Accessibility

### CODE REVIEW

**Comment by momesgin** [comment]
@richard-cox the failed e2e test related to configmaps is not related to the changes of this PR. As I've mentioned in the "Areas to test..." section of the PR's description I can't find a real example to test my changes other than for the vSphere addons, but I think I've covered all the cases in the unit tests. Please let me know if you're aware of any other examples so I can test them myself as well.

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 1]
```diff
@@ -921,14 +922,17 @@ export default {
       }
     },
 
-    selectedVersion() {
-      this.versionInfo = {}; // Invalidate cache such that version info relevent to selected kube version is updated
+    selectedVersion: {
+      async handler(neu, old) {
+        await this.preserveAddonConfigs(old, neu);
+        this.versionInfo = {}; // Invalidate cache such that version info relevent to selected kube version is updated
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]:
```suggestion
        this.versionInfo = {}; // Invalidate cache such that version info relevant to selected kube version is updated
```

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 2]
```diff
@@ -972,6 +976,123 @@ export default {
...
 
   methods: {
+    /**
+     * When the Kubernetes version is changed, this method is called to handle the add-on configurations
+     * for all enabled addons. It checks if an addon's version has changed and, if so, determines if the
+     * user's custom configurations should be preserved for the new version.
+     *
+     * The goal is to avoid showing a confirmation dialog for changes in default values that the user has not customized.
+     */
+    async preserveAddonConfigs(oldVersion, newVersion) {
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: to avoid method count creep in this file again can these three nicely succinct methods be moved out to a mixin, or i guess nowadays it's setup calling a composable that supplies the three methods?

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 3]
```diff
@@ -972,6 +976,123 @@ export default {
...
+    },
+
+    /**
+     * Processes a single add-on version change. It fetches the old and new chart information,
+     * calculates the differences in default values, and filters them based on user's customizations.
+     * If there are no significant differences, it preserves the user's custom values for the new version.
+     */
+    async _processAddonVersionChange(chartName, oldAddon, newAddon) {
+      try {
+        const [oldVersionInfo, newVersionInfo] = await Promise.all([
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: initAddons, that runs after preserveAddonConfigs in the watch on selectedVersion, has some safe guards against  charts with a name that includes `none`. guessing we need this here too

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 4]
```diff
@@ -972,6 +976,123 @@ export default {
...
+            repoType:    'cluster',
+            repoName:    newAddon.repo,
+            chartName,
+            versionName: newAddon.version,
+          })
+        ]);
+
+        const oldDefaults = oldVersionInfo.values;
+        const newDefaults = newVersionInfo.values;
+        const differences = diff(oldDefaults, newDefaults);
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: super nitpic
- could this be called something like `defaultsDifferences`
- then `finalDifferences` something like defaultsAndUserDifferences

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 5]
```diff
@@ -972,6 +976,123 @@ export default {
...
+
+        const userOldValues = this.userChartValues[`${ chartName }-${ oldAddon.version }`];
+
+        // We only care about differences in values that the user has actually customized.
+        // If the user hasn't touched a value, a change in its default should not be considered a breaking change.
+        const finalDifferences = userOldValues ? this._filterRelevantDifferences(differences, userOldValues) : {};
+
+        this.addonConfigDiffs[chartName] = finalDifferences;
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: This could be

```
this.addonConfigDiffs[chartName] = false;

then inside `if` below 

this.addonConfigDiffs[chartName] = true

```

OR

```
this.addonConfigDiffs[chartName] = !isEmpty(finalDifferences)
if !this.addonConfigDiffs[chartName] {....
```

in both cases the final check is neater

`const hasDiffs = Object.values(this.addonConfigDiffs).some((d) => !isEmpty(d));` 
is neater with 
`const hasDiffs = Object.values(this.addonConfigDiffs).includes(true);`

File: shell/edit/provisioning.cattle.io.cluster/__tests__/rke2.test.ts [hunk 6]
```diff
@@ -610,4 +612,215 @@ describe('component: rke2', () => {
 
     expect(wrapper.vm.value.spec.rkeConfig.chartValues).toStrictEqual(expected);
   });
+
+  describe('preserveAddonConfigs', () => {
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/__tests__/rke2.test.ts]: this is a good set of tests, but given the fun we've had with diffs and addons can we chuck a lot more edge cases at it>


properties that are
- objects
- arrays
- arrays of objects

covering diffs like
- items added, changed or removed in an array
- properties that have been added or removed

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: now we know that the user has edits which clash with the values we're going to use... that will ultimately be ignored... which should warn them on the `Add-on: x` tab instead of show the generic `Add-On Configuration can vary between Kubernetes versions. Changing the Kubernetes version may reset the values below.` when kube has changed.

Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: when there are clashes, where did we get to with showing the old config alongside the current defaults?

Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: This also means the modal text shown just before submit can change from `Changing the Kubernetes Version can reset Add-On values. You should check that the values are as expected before continuing.`

to something like

`Changing the Kubernetes Version has provided new values for Add-On x,y and z which clash with your provided values. As a result please re-apply them before continuing.`

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 1 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: done

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 2 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: I moved them to `shell/utils/cluster.js` instead, since `rke2.vue` uses the old options API, and mixins are really unlikeable :D

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 3 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: thanks for catching this, done

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 4 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: much better naming! done

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: I did this refactor, but then reverted it because I thought maybe we can display this data on the modal? I'll go back to boolean if we decide not to

File: shell/edit/provisioning.cattle.io.cluster/__tests__/rke2.test.ts (hunk 6 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/__tests__/rke2.test.ts]: yes you're absolutely right, I'll come back to the tests when we're happy with the final code

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: > now we know that the user has edits which clash with the values we're going to use... that will ultimately be ignored... which should warn them on the Add-on: x tab instead of show the generic Add-On Configuration can vary between Kubernetes versions. Changing the Kubernetes version may reset the values below. when kube has changed.

Just to double-check, now we'll see the warning banner only on the tabs that are actually affected by the kube version change, is that what we want?

Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: > This also means the modal text shown just before submit can change from `Changing the Kubernetes Version can reset Add-On values. You should check that the values are as expected before continuing.`
> 
> to something like
> 
> `Changing the Kubernetes Version has provided new values for Add-On x,y and z which clash with your provided values. As a result please re-apply them before continuing.`

That's a great idea. Not sur if we want to display a more detailed message like mentioning the exact properties for each addon, if not then I'll update the code to return boolean instead of object as you mentioned in your comment before.

Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: We haven't followed up yet, hopefully we improve this in 2.14.

I'm also wondering if we should apply these new changes to the create mode too?

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 2 above)
Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: composables at their core are just functions that bring in properties and functions. so something like this could have been used

```
data() {
  const { preserveAddonConfigs } = useV2ProvAddonConfig();
  return {
    preserveAddonConfigs
  }
}
```

though i think options api still supports `setup`

```
setup() {
  const { preserveAddonConfigs } = useV2ProvAddonConfig();
  return {
    preserveAddonConfigs
  }
}
```

in the interests of time though lets just keep it in that file

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 7]
```diff
@@ -921,14 +923,17 @@ export default {
       }
     },
 
-    selectedVersion() {
-      this.versionInfo = {}; // Invalidate cache such that version info relevent to selected kube version is updated
+    selectedVersion: {
+      async handler(neu, old) {
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: this doesn't need a handler, the async can just be directly on `selectedVersion`

File: shell/utils/cluster.js [hunk 8]
```diff
@@ -335,3 +335,132 @@ export async function initSchedulingCustomization(value, features, store, mode)
...
+
+/**
+ * Recursively filters a `diffs` object to only include differences that are relevant to the user.
+ * A difference is considered relevant if the user has provided a custom value for that specific field.
+ *
+ * @param {object} diffs - The object representing the differences between two chart versions' default values.
+ * @param {object} userVals - The object containing the user's custom values.
+ * @returns {object} A new object containing only the relevant differences.
+ */
+function _filterRelevantDifferences(diffs, userVals) {
```
Comment (richard-cox) [inline shell/utils/cluster.js]: given these three functions aren't nicely grouped in a file / function can the name's be prefixed to make it clear they're all 1-1 connected?

_filterRelevantDifferences
_processAddonVersionChange
preserveAddonConfigs

becomes

_addonConfigPreserveFilter
_addonConfigPreserveProcess
addonConfigPreserve

File: shell/utils/cluster.js [hunk 9]
```diff
@@ -335,3 +335,132 @@ export async function initSchedulingCustomization(value, features, store, mode)
...
+}
+
+/**
+ * When the Kubernetes version is changed, this method is called to handle the add-on configurations
+ * for all enabled addons. It checks if an addon's version has changed and, if so, determines if the
+ * user's custom configurations should be preserved for the new version.
+ *
+ * The goal is to avoid showing a confirmation dialog for changes in default values that the user has not customized.
+ */
+export async function preserveAddonConfigs(component, oldVersion, newVersion) {
```
Comment (richard-cox) [inline shell/utils/cluster.js]: jsdoc needs arg descriptions

Comment (richard-cox) [inline shell/utils/cluster.js]: it's best to pass specific required values to helper functions instead of large objects with additional properties that aren't needed. this is especially the case for components (if the caller component changes it could break obscure references in helper functions)

instead only those values that are needed should be specified. 

```suggestion
export async function preserveAddonConfigs({ isEdit, addonConfigDiffs, addonNames, etc... }, oldVersion, newVersion) {
```

then ideally explicitly passed through

```
const { isEdit, addonConfigDiffs, addonNames, etc...} = this;
await preserveAddonConfigs({ isEdit, addonConfigDiffs, addonNames, etc...} , old, neu);
```

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: `Add-on: <name>` tab
- we now know that this specific config has merged issues, so we need to show a message with a good level of information for the user to understand what's happened/happening
- only adding the message if there is a merge issue is good
- we now need to change the generic statement to something more specific
- `Add-On Configuration can vary between Kubernetes versions. Changing the Kubernetes version may reset the values below.` --> `Add-On Configuration has changed between Kubernetes versions <previous> and <new> and conflict with your previous values. Please reapply them`
- YAML based forms - we show that second yaml block `Previous Values` on the right containing the users previous values
- Questions based forms - if a solution still hasn't been determined can you convert the original issue to an epic, create a step 1 issue covering the changes in this PR and a step 2 issue covering the questions based form changes?


Final Warning Model
- we look again at everything given possible changes
  - user has edited the yaml of an addon that has a conflict - ignore addon
  - user has not edited the yaml of an addon that has a conflict - track
- if there are tracked addons we include them in the message to the user
  - `Changing the Kubernetes Version can reset Add-On values. You should check that the values are as expected before continuing.` --> `Add-On Configuration has changed between Kubernetes versions <previous> and <new> and conflict with your previous values for Add-On x,y and z. As a result please re-apply them before continuing.`

Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: > when there are clashes, where did we get to with showing the old config alongside the current defaults?

now covered by https://github.com/rancher/dashboard/pull/15445#discussion_r2426579565

> I'm also wondering if we should apply these new changes to the create mode too?

there can't be any conflicts there, as there's no previous to new addon conflicts, or previous user values

**Review by richard-cox — CHANGES_REQUESTED** [review]

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: > there can't be any conflicts there, as there's no previous to new addon conflicts, or previous user values

it's really an edge case, but while you're creating the cluster you can make changes to the addon configs and then change the kube version

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 7 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: done

File: shell/utils/cluster.js (hunk 8 above)
Comment (momesgin) [reply shell/utils/cluster.js]: done

File: shell/utils/cluster.js (hunk 9 above)
Comment (momesgin) [reply shell/utils/cluster.js]: done

Comment (momesgin) [reply shell/utils/cluster.js]: done

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: > YAML based forms - we show that second yaml block Previous Values on the right containing the users previous values

I thought we don't display the diff until we have a solution for both YAML and form cases?

> user has edited the yaml of an addon that has a conflict - ignore addon

What's the definition of "edited" in this context? **any** changes made to the addon that has a conflict? if yes, I'm thinking of adding a warning/error icon next to each addon tab that has a conflict(we might need to make some small changes to the tab component as currently it has a hardcoded tooltip for that icon), and then removing the conflict for each tab as soon as the user changes the addon configuration. By doing this we don't need to re-run the calculation on the confirm modal since the conflict tracking's already been updated.

Also we can remove the warning banner and display it as a tooltip on the tab icon instead.

Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: Update: I realized I was comparing any previously selected k8s version vs the newly selected one, but I should've always compared the original k8s version against the newly selected version. So now I think it would be better not to make things more complicated and stick to edit mode only.

Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: @oboc-sts @richard-cox to make it clear I've recorded a video to show my suggestion(except I kept the warning banner and didn't update the error tooltip). Note that since I was not able to find a real case where the default values being changed for a form based case like vSphere CPI I had to mock some data to be able to get the error for it.

https://github.com/user-attachments/assets/9ad454b5-18e3-4391-bf69-7e6053c49a4c

Comment (oboc-sts) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: I can't spend too much time on this, but from the looks of it... I'd say we could pursue this path as an improvement.

File: shell/assets/translations/en-us.yaml [hunk 10]
```diff
@@ -472,10 +472,7 @@ addClusterMemberDialog:
 
 addonConfigConfirmation:
   title: Add-On Reset
-  body: Changing the Kubernetes Version can reset Add-On values. You should check that the values are as expected before continuing.
-
-addProjectMemberDialog:
```
Comment (richard-cox) [inline shell/assets/translations/en-us.yaml:477]: we need to keep this

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 11]
```diff
@@ -921,8 +928,24 @@ export default {
...
+        const { addonConfigDiffs, addonNames, userChartValues } = this;
+        const context = {
+          addonConfigDiffs,
+          addonNames,
+          $store: this.$store,
+          userChartValues
+        };
+
+        await addonConfigPreserve(
+          context,
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]: This can be simplified to 
```
       const {
          addonConfigDiffs, addonNames, userChartValues, $store
        } = this;

        await addonConfigPreserve(
          {
            addonConfigDiffs, addonNames, userChartValues, $store
          },
          this.originalKubeVersion?.charts,
          neu?.charts
        );
```

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue [hunk 12]
```diff
@@ -921,8 +928,24 @@ export default {
       }
     },
 
-    selectedVersion() {
-      this.versionInfo = {}; // Invalidate cache such that version info relevent to selected kube version is updated
+    async selectedVersion(neu, old) {
```
Comment (richard-cox) [inline shell/edit/provisioning.cattle.io.cluster/rke2.vue]:
```suggestion
    async selectedVersion(neu) {
```

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: >> YAML based forms - we show that second yaml block Previous Values on the right containing the users previous values

> I thought we don't display the diff until we have a solution for both YAML and form cases?

Part of me thinks this is such a small change (additional YamlEditor in read only mode) and as it would make a huge difference that we go ahead.

>> user has edited the yaml of an addon that has a conflict - ignore addon

> What's the definition of "edited" in this context? 

my thinking is if the user HAS re-applied their values and then hit save again they still see the same `please re-apply them before continuing` message. in this world we can assume if they've changed something they're happy with the config for that addon.

Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: given we're sticking with edit, will close this one

File: shell/assets/translations/en-us.yaml (hunk 10 above)
Comment (momesgin) [reply shell/assets/translations/en-us.yaml:477]: ahhh must have mistakenly selected the extra lines

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: > Part of me thinks this is such a small change (additional YamlEditor in read only mode) and as it would make a huge difference that we go ahead.

I haven't really looked into how much work it will add, but I'm worried about starting it and then new ux/ui questions would come around it. Our communications on the ux/ui of this task has been quite messy, and I'm still not really sure what the final decision is. Should I go with my screen recording approach? should we remove warning banner and show tooltip instead or both? everyone's calendar is packed with meetings, could you please schedule a meeting for us and Ken to finalize this?

Comment (richard-cox) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: Had a chat with Neil, we think we could probably close the related JIRAs with what's in the PR atm.

Going forward though I think there's still a lot of major frustrations we can avert by adding the previous user values to at least the YAML page, even if in a basic way. At the moment the user has a lot of convoluted steps when we could easily show something in that context

- find the previous values by either
  - click cancel --> edit --> addon tab
  - go back to previous kube version
- copy their addon values to an external diff tool
- reset the state to get the conflict again on the addon tab
- copy the new addon default values to an external diff tool
- work out the diff in that tool and then copy over a new version

Can you create an issue to improve the experience of re-applying helm values to addon configs, and also one to show an additional tab warning for addons with conflicts?

Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: Sounds good, will take care of the new issues

File: shell/edit/provisioning.cattle.io.cluster/__tests__/rke2.test.ts [hunk 13]
```diff
@@ -475,16 +475,17 @@ describe('component: rke2', () => {
         },
         provider: 'custom'
       },
-      data: () => ({
-        agentArgs: {
+      computed: {
+        ...rke2.computed,
```
Comment (momesgin) [inline shell/edit/provisioning.cattle.io.cluster/__tests__/rke2.test.ts:479]: kept these changes, it's for getting rid of console warnings

File: shell/edit/provisioning.cattle.io.cluster/rke2.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/provisioning.cattle.io.cluster/rke2.vue]: New issues are created:
https://github.com/rancher/dashboard/issues/15688
https://github.com/rancher/dashboard/issues/15687

We'll follow-up with the design team to add more details to the issues
